### ISSUE TITLE
Non-intuitive behavior of bitwise operations compiled into JavaScript

### ISSUE THREAD

**Issue by BlueRayi** [dart-lang/language#2354]
TL;DR
Currently, Dart compiled into JavaScript considers the operands of bitwise operations as unsigned 32-bit integers.

I wrote and ran the following code

```dart
void main() {
  print(-1 & -1);
  print(-2 ^ 0);
  print(-3 | 0);
  print(~~-4);
  print(-5 << 0);
  print(-6 >> 0);
}
```

`&-1`, `^0`, `|0`, `~~`, `<<0`, `>>0`
These are all known to be identity functions for integers. So, of course, this is what happens when you perform this.

```
-1
-2
-3
-4
-5
-6
```

However, when we compile this into JavaScript code and look at the console, something strange is happening.

```
4294967295
4294967294
4294967293
4294967292
4294967291
4294967290
```

This is obviously happening because negative values (expressed in 2's complement) are interpreted as unsigned integers. It is my opinion that they should be considered signed 32-bit integers.

In particular, the [reference](https://api.dart.dev/dev/2.19.0-20.0.dev/dart-core/int/operator_bitwise_negate.html) for `~` reads as follows:

> This maps any integer x to -x - 1.

Therefore, I consider this implementation inappropriate.

Postscript.

This is not directly related to the above, but the fact that `>>>0` is a identity function even for negative values is different from JavaScript semantics and confuses me a bit.

**Comment by julemand101** [comment]
Looks like the issue should have been posted in the SDK repository: https://github.com/dart-lang/sdk/issues

This issue tracker are for the Dart language specification and this seems more related to the implementation of the Dart language than a specification issue.
